File: app/app4_processor.js
import fs from 'fs';
import path from 'path';
import ExcelJS from 'exceljs';
import JSZip from 'jszip';
import stringSimilarity from 'string-similarity';

function cleanString(val) {
    if (val === null || val === undefined) return "";
    if (typeof val === 'number') return val.toString();
    if (typeof val === 'object' && val.text) return String(val.text).trim(); // Handle rich text
    if (val instanceof Date) return val.getFullYear().toString();
    return String(val).trim().replace(/\.0$/, '');
}

function normalizeForComparison(val) {
    if (!val) return "";
    return String(val).toUpperCase().replace(/\s+/g, '').trim();
}

function extractRoomCode(val) {
    const s = normalizeForComparison(val);
    const match = s.match(/^([A-Z0-9]+)/);
    return match ? match[1] : s;
}

function extractCleanRoomName(val) {
    const norm = normalizeForComparison(val);
    return norm.replace(/-\d{7,}.*$/, '').trim();
}

function isSameRoom(opname, master) {
    const normOpname = normalizeForComparison(opname);
    const normMaster = normalizeForComparison(master);
    if (!normOpname || !normMaster) return false;
    if (normOpname === normMaster) return true;
    
    if (extractRoomCode(opname) !== extractRoomCode(master)) return false;
    
    const cleanOpname = extractCleanRoomName(opname);
    const cleanMaster = extractCleanRoomName(master);
    if (cleanOpname === cleanMaster) return true;
    if (cleanMaster.includes(cleanOpname) || cleanOpname.includes(cleanMaster)) return true;
    if (Math.min(cleanOpname.length, cleanMaster.length) <= 22) return false;
    
    return stringSimilarity.compareTwoStrings(cleanOpname, cleanMaster) >= 0.68;
}

function isNonNumericOracleCode(val) {
    if (!val || val === '-' || val === 'nan' || val === 'None' || val === '') return false;
    return !/^\d+$/.test(val.trim());
}

function isEmptyVal(val) {
    if (!val) return true;
    const s = String(val).trim();
    return s === '' || s === '-' || s === 'nan' || s === 'None' || s === 'null';
}

async function loadAspxDict(aspxPath) {
    if (!aspxPath || !fs.existsSync(aspxPath)) return {};
    const wb = new ExcelJS.Workbook();
    await wb.xlsx.readFile(aspxPath);
    const ws = wb.worksheets[0];
    if (!ws) return {};

    const headers = {};
    ws.getRow(1).eachCell((cell, colNum) => headers[cleanString(cell.value).toUpperCase().trim()] = colNum);

    const aspxDict = {};
    ws.eachRow((row, rowNumber) => {
        if (rowNumber === 1) return;
        const barcode = cleanString(row.getCell(headers['BARCODE']).value).toUpperCase();
        if (isEmptyVal(barcode)) return;

        const getValue = (colName) => {
            const raw = headers[colName] ? cleanString(row.getCell(headers[colName]).value) : '';
            return (!isEmptyVal(raw) && raw.toUpperCase() !== 'NULL') ? raw : '';
        };

        aspxDict[barcode] = {
            oracleId: getValue('ORACLE ID'),
            noPo: getValue('NO PO'),
            bln: getValue('BLN'),
            thn: getValue('THN')
        };
    });
    return aspxDict;
}

function resolveAssetOracle(masterOracleVal, barcode, aspxDict) {
    const aspxEntry = aspxDict[barcode];
    if (isEmptyVal(masterOracleVal)) return aspxEntry?.oracleId || '-';
    if (isNonNumericOracleCode(masterOracleVal)) return aspxEntry?.oracleId || masterOracleVal;
    return masterOracleVal;
}

function resolveTahunPerolehan(masterTahunVal, barcode, aspxDict) {
    if (!isEmptyVal(masterTahunVal) && masterTahunVal !== '-') return masterTahunVal;
    const aspxEntry = aspxDict[barcode];
    if (aspxEntry) {
        if (aspxEntry.thn && aspxEntry.bln) return `${String(aspxEntry.bln).padStart(2, '0')}/${aspxEntry.thn}`;
        return aspxEntry.thn || aspxEntry.bln || '-';
    }
    return '-';
}

function resolveKondisiMaster(masterKondisiVal, opnameKeterangan) {
    if (!isEmptyVal(masterKondisiVal) && masterKondisiVal !== '-') return masterKondisiVal;
    const catatan = cleanString(opnameKeterangan).trim();
    return (!catatan || catatan === '-') ? 'BAIK' : '-';
}

async function loadMasterData(masterPath) {
    const wb = new ExcelJS.Workbook();
    await wb.xlsx.readFile(masterPath);
    const ws = wb.worksheets[0];

    const headers = {};
    ws.getRow(1).eachCell((cell, colNum) => headers[cleanString(cell.value)] = colNum);

    const dict = {};
    ws.eachRow((row, rowNumber) => {
        if (rowNumber === 1) return;
        const barcode = cleanString(row.getCell(headers['NO BARCODE']).value).toUpperCase();
        if (barcode && barcode !== '-' && barcode !== 'NONE' && barcode !== 'NAN') {
            const getVal = (key) => headers[key] ? cleanString(row.getCell(headers[key]).value) : '-';
            dict[barcode] = {
                'NO BARCODE': getVal('NO BARCODE'),
                'ASSET ORACLE': getVal('ASSET ORACLE'),
                'LOKASI': getVal('LOKASI'),
                'JENIS HARTA': getVal('JENIS HARTA'),
                'KONDISI': getVal('KONDISI'),
                'TAHUN Perolehan': getVal('TAHUN Perolehan') || getVal('Tahun Perolehan') || getVal('TAHUN PEROLEHAN') || '-'
            };
        }
    });
    return dict;
}

function formatRecouncilSheet(ws) {
    for (let i = 0; i < 7; i++) ws.getColumn(11 + i).hidden = false;
    
    const headers = ['NO BARCODE', 'ASSET ORACLE', 'LOKASI', 'JENIS HARTA', 'KONDISI MASTER', 'TAHUN Perolehan', 'HASIL RECOUNCIL'];
    headers.forEach((h, i) => {
        const cell = ws.getCell(7, 11 + i);
        cell.value = h;
        cell.font = { name: 'Calibri', size: 11, bold: true, color: { argb: 'FF000000' } };
        cell.border = { top: { style: 'thin' }, left: { style: 'thin' }, bottom: { style: 'thin' }, right: { style: 'thin' } };
        cell.alignment = { vertical: 'middle', horizontal: 'center', wrapText: true };
        cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: i < 6 ? 'FFD9D9D9' : 'FF00B0F0' } };
    });
}

async function runRecouncil(opnames, masterPath, outputPath, aspxPath) {
    try {
        if (!fs.existsSync(masterPath)) return { status: 'error', message: 'File Master Data tidak ditemukan.' };

        const masterDict = await loadMasterData(masterPath);
        const aspxDict = await loadAspxDict(aspxPath);
        const processedFiles = [];
        let firstSuggestedName = null;

        for (const opname of opnames) {
            const wb = new ExcelJS.Workbook();
            await wb.xlsx.readFile(opname.path);
            const ws = wb.getWorksheet('Recouncil') || wb.worksheets[0];

            let roomNameFromSheet = "";
            const row2CellA = ws.getCell('A2').value;
            const textVal = (typeof row2CellA === 'object' && row2CellA?.richText)
                ? row2CellA.richText.map(rt => rt.text).join('') : String(row2CellA || '');

            if (textVal.toUpperCase().includes('RUANGAN')) {
                roomNameFromSheet = textVal.replace(/RUANGAN\s*:\s*/i, '').trim().replace(/[\\/:*?"<>|]/g, '-');
            }

            const finalFileName = roomNameFromSheet ? `${roomNameFromSheet}.xlsx` : `${opname.original.replace(/\.[^/.]+$/, "")}.xlsx`;
            if (!firstSuggestedName) firstSuggestedName = finalFileName;

            let lastDataRow = Math.max(ws.rowCount, 8);
            for (let r = ws.rowCount; r > 7; r--) {
                if ([1,2,3,4,5,6,7,8,9,10].some(c => cleanString(ws.getCell(r, c).value) !== "")) {
                    lastDataRow = r; break;
                }
            }

            formatRecouncilSheet(ws);

            for (let r = 8; r <= lastDataRow; r++) {
                const rawBarcode = cleanString(ws.getCell(r, 3).value).toUpperCase();
                const ruanganOpname = cleanString(ws.getCell(r, 6).value);
                const kolomI = ws.getCell(r, 9).value;

                let statusRecouncil = "Barcode Kosong di Opname";
                let outMaster = { 'NO BARCODE': '-', 'ASSET ORACLE': '-', 'LOKASI': '-', 'JENIS HARTA': '-', 'KONDISI MASTER': '-', 'TAHUN Perolehan': '-' };

                if (rawBarcode && rawBarcode !== 'nan' && rawBarcode !== '-') {
                    if (!masterDict[rawBarcode]) {
                        statusRecouncil = "Barcode Belum Sesuai, Asset di Oracle tidak ada";
                        outMaster['NO BARCODE'] = rawBarcode;
                        outMaster['ASSET ORACLE'] = resolveAssetOracle('-', rawBarcode, aspxDict);
                    } else {
                        const mData = masterDict[rawBarcode];
                        outMaster = {
                            'NO BARCODE': rawBarcode,
                            'ASSET ORACLE': resolveAssetOracle(mData['ASSET ORACLE'], rawBarcode, aspxDict),
                            'LOKASI': mData['LOKASI'],
                            'JENIS HARTA': mData['JENIS HARTA'],
                            'KONDISI MASTER': resolveKondisiMaster(mData['KONDISI'], kolomI),
                            'TAHUN Perolehan': resolveTahunPerolehan(mData['TAHUN Perolehan'], rawBarcode, aspxDict)
                        };

                        if (isEmptyVal(outMaster['JENIS HARTA']) || outMaster['JENIS HARTA'] === '-') {
                            statusRecouncil = "Data Ditemukan Tidak Lengkap";
                        } else if (ruanganOpname && !isSameRoom(ruanganOpname, outMaster['LOKASI'])) {
                            statusRecouncil = `Salah Ruangan, hasil ruangan opname berada di: ${ruanganOpname}`;
                        } else {
                            statusRecouncil = "Sudah Sesuai";
                            outMaster['LOKASI'] = ruanganOpname;
                        }
                    }
                }

                [outMaster['NO BARCODE'], outMaster['ASSET ORACLE'], outMaster['LOKASI'], outMaster['JENIS HARTA'], outMaster['KONDISI MASTER'], outMaster['TAHUN Perolehan'], statusRecouncil].forEach((val, i) => {
                    const cell = ws.getCell(r, 11 + i);
                    cell.value = val;
                    cell.font = { name: 'Calibri', size: 11 };
                    cell.border = { top: { style: 'thin' }, left: { style: 'thin' }, bottom: { style: 'thin' }, right: { style: 'thin' } };
                    cell.alignment = { vertical: 'middle', horizontal: 'center', wrapText: true };
                });
            }

            // Fixed auto-width formatting
            for (let col = 1; col <= 17; col++) {
                ws.getColumn(col).width = [3, 11, 11, 12, 13, 14].includes(col) ? 20 : 15;
            }

            if (opnames.length === 1) {
                await wb.xlsx.writeFile(outputPath);
                processedFiles.push(outputPath);
            } else {
                const tempOut = path.join(path.dirname(outputPath), `Temp_${Date.now()}_${path.basename(opname.path)}`);
                await wb.xlsx.writeFile(tempOut);
                processedFiles.push({ path: tempOut, finalName: finalFileName });
            }
        }

        if (opnames.length > 1) {
            const zip = new JSZip();
            for (const pf of processedFiles) {
                zip.file(pf.finalName, fs.readFileSync(pf.path));
                fs.unlinkSync(pf.path);
            }
            fs.writeFileSync(outputPath, await zip.generateAsync({ type: 'nodebuffer', compression: 'DEFLATE' }));
        }

        return { status: "success", message: "Proses berhasil", output: outputPath, files_processed: opnames.length, suggested_filename: opnames.length === 1 ? firstSuggestedName : `Recouncil_Result_${Date.now()}.zip` };
    } catch (err) {
        console.error("Error in recouncil processing:", err);
        return { status: "error", message: err.message, traceback: err.stack };
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length < 2) process.exit(1);
    if (args[0] === "process") {
        try {
            const data = JSON.parse(fs.readFileSync(args[1], 'utf8'));
            console.log(JSON.stringify(await runRecouncil(data.opnames, data.master, data.output_path, data.aspx || null)));
        } catch (e) {
            console.log(JSON.stringify({ status: "error", message: String(e) }));
        }
    }
}

main();

File: app/app4_processor.py
import pandas as pd
import numpy as np
import os
import re
import traceback
import json
import zipfile
import shutil
from openpyxl import load_workbook
from openpyxl.styles import PatternFill, Font, Border, Side, Alignment
from openpyxl.utils import get_column_letter

DEFAULT_FONT_EXCEL = Font(name='Calibri', size=11)
HEADER_FONT_EXCEL = Font(name='Calibri', size=11, bold=True, color='000000')
THIN_BORDER = Border(left=Side(style='thin'), right=Side(style='thin'), top=Side(style='thin'), bottom=Side(style='thin'))
CENTER_ALIGNMENT = Alignment(horizontal='center', vertical='center', wrap_text=True)

MASTER_FILL = PatternFill(start_color='FFC000', end_color='FFC000', fill_type='solid') # Kuning/orange untuk data Oracle MS
RESULT_FILL = PatternFill(start_color='00B0F0', end_color='00B0F0', fill_type='solid') # Biru cerah (hasil recouncil)

def clean_string(val):
    if pd.isna(val) or val is None:
        return ""
    if isinstance(val, float):
        if val.is_integer():
            return str(int(val))
        return str(val).rstrip('0').rstrip('.')
    if isinstance(val, int):
        return str(val)
    if isinstance(val, str):
        val = val.strip()
        if val.endswith('.0'):
            val = val[:-2]
        return val
    return str(val).strip()

def make_unique_cols(cols):
    seen = {}
    new_cols = []
    for c in cols:
        new_c = str(c).strip()
        if new_c in seen:
            seen[new_c] += 1
            new_cols.append(f"{new_c}_{seen[new_c]}")
        else:
            seen[new_c] = 0
            new_cols.append(new_c)
    return new_cols

def style_cell(cell, is_header=False, section="master"):
    cell.font = HEADER_FONT_EXCEL if is_header else DEFAULT_FONT_EXCEL
    cell.border = THIN_BORDER
    cell.alignment = CENTER_ALIGNMENT
    if is_header:
        if section == "master":
            cell.fill = MASTER_FILL
        else:
            cell.fill = RESULT_FILL

def run_recouncil(opname_paths, master_path, output_path):
    """
    Memproses data Opname (App2) dengan Data Master (App3).
    Mengedit file Opname in-place untuk mempertahankan format aslinya (headers, judul, dsb),
    hanya menambahkan kolom K-Q di sebelah kanan tabel aslinya.
    Jika multiple file, mengembalikan bentuk zip.
    """
    if not isinstance(opname_paths, list):
        opname_paths = [opname_paths]
        
    for path in opname_paths:
        if not os.path.exists(path):
            return {"status": "error", "message": f"File sumber tidak ditemukan: {path}"}
            
    if not os.path.exists(master_path):
        return {"status": "error", "message": "File Master Data tidak ditemukan."}

    try:
        # 1. Load Master Data ke Dictionary
        df_master = pd.read_excel(master_path)
        df_master.columns = make_unique_cols(df_master.columns)
        
        master_dict = {}
        for _, row in df_master.iterrows():
            barcode = clean_string(row.get('NO BARCODE', ''))
            if barcode and barcode not in ['nan', 'None', '', '-']:
                master_data = {
                    'NO BARCODE': clean_string(row.get('NO BARCODE', '-')),
                    'ASSET ORACLE': clean_string(row.get('ASSET ORACLE', '-')),
                    'LOKASI': clean_string(row.get('LOKASI', '-')),
                    'JENIS HARTA': clean_string(row.get('JENIS HARTA', '-')),
                    'KONDISI': clean_string(row.get('KONDISI', '-')),
                    'TAHUN Perolehan': clean_string(row.get('Tahun Perolehan', '-'))
                }
                
                # Handle dates for excel writing
                tp = master_data['TAHUN Perolehan']
                if hasattr(tp, 'date'):
                    master_data['TAHUN Perolehan'] = tp.strftime('%Y')
                elif isinstance(tp, str) and ' ' in tp and ':' in tp:
                     master_data['TAHUN Perolehan'] = tp.split(' ')[0][:4]
                     
                master_dict[barcode] = master_data

        processed_files = []
        
        # 2. Proses masing-masing Opname Path secara In-Place
        for opname_path in opname_paths:
            wb = load_workbook(opname_path)
            
            # Cari sheet yang digunakan (prioritas: Recouncil)
            ws = wb['Recouncil'] if 'Recouncil' in wb.sheetnames else wb.active
                
            # Deteksi baris terakhir yang memiliki data di kolom A-J supaya tidak format sampai ribuan ke bawah
            last_data_row = 7
            for r in range(ws.max_row, 7, -1):
                has_data = False
                for c in range(1, 11): # Cek kolom A sampai J
                    val = ws.cell(row=r, column=c).value
                    if val is not None and str(val).strip() != "":
                        has_data = True
                        break
                if has_data:
                    last_data_row = r
                    break
            
            # Jika last_data_row mentok di 7 dan max_row jauh lebih besar, bisa jadi sheet kosong
            if last_data_row < 8:
                last_data_row = max(ws.max_row, 8)
                
            # Kita tulis ulang header di baris 7 agar formatnya rapi
            new_headers = ['NO BARCODE', 'ASSET ORACLE', 'LOKASI', 'JENIS HARTA', 'KONDISI MASTER', 'TAHUN Perolehan', 'HASIL RECOUNCIL']
            for i, h_name in enumerate(new_headers):
                cell = ws.cell(row=7, column=11 + i) # Mulai dari Kolom K (11)
                cell.value = h_name
                style_cell(cell, is_header=True, section="master" if i < 6 else "result")

            # Mulai memproses data tepat di bawah header (baris 8 ke bawah)
            for r in range(8, last_data_row + 1):
                raw_barcode = ws.cell(row=r, column=3).value # Pasti Kolom C (Index 3)
                raw_ruang = ws.cell(row=r, column=6).value   # Pasti Kolom F (Index 6)
                
                barcode_opname = clean_string(raw_barcode)
                ruangan_opname = clean_string(raw_ruang)
                
                status_recouncil = ""
                out_master = {
                    'NO BARCODE': '-', 'ASSET ORACLE': '-', 'LOKASI': '-', 
                    'JENIS HARTA': '-', 'KONDISI MASTER': '-', 'TAHUN Perolehan': '-'
                }
                
                if not barcode_opname or barcode_opname in ['nan', '-', 'None', '']:
                    status_recouncil = "Barcode Kosong di Opname"
                elif barcode_opname not in master_dict:
                    status_recouncil = "Barcode Belum Sesuai, Asset di Oracle tidak ada"
                    out_master['NO BARCODE'] = barcode_opname # Force paste barcode opname here
                else:
                    m_data = master_dict[barcode_opname]
                    out_master = {
                        'NO BARCODE': barcode_opname, # Always use the scanned barcode
                        'ASSET ORACLE': m_data['ASSET ORACLE'],
                        'LOKASI': m_data['LOKASI'],
                        'JENIS HARTA': m_data['JENIS HARTA'],
                        'KONDISI MASTER': m_data['KONDISI'],
                        'TAHUN Perolehan': m_data['TAHUN Perolehan']
                    }
                    
                    if out_master['JENIS HARTA'] in ['', '-', 'nan', 'None'] or out_master['TAHUN Perolehan'] in ['', '-', 'nan', 'None']:
                        status_recouncil = "Data Ditemukan Tidak Lengkap"
                    elif ruangan_opname.upper() != out_master['LOKASI'].upper() and ruangan_opname:
                        status_recouncil = f"Salah Ruangan seharusnya di {out_master['LOKASI']}"
                    else:
                        status_recouncil = "Sudah Sesuai"
                        
                # Tulis Master Data dan Hasil ke Excel Document (Kolom K sampai Q)
                master_vals = [
                    out_master['NO BARCODE'], out_master['ASSET ORACLE'], out_master['LOKASI'], 
                    out_master['JENIS HARTA'], out_master['KONDISI MASTER'], out_master['TAHUN Perolehan']
                ]
                
                for i, m_val in enumerate(master_vals):
                    cell = ws.cell(row=r, column=11 + i)
                    cell.value = m_val
                    style_cell(cell, is_header=False)
                    
                q_cell = ws.cell(row=r, column=17)
                q_cell.value = status_recouncil
                style_cell(q_cell, is_header=False)

            # Auto Column Width for K-Q
            widths = [15, 15, 30, 30, 10, 15, 35]
            for i, w in enumerate(widths):
                col_letter = get_column_letter(11 + i)
                ws.column_dimensions[col_letter].width = w

            # Save modified file to output path
            base_fname = os.path.basename(opname_path)
            if len(opname_paths) == 1:
                # Single file -> Directly to requested .xlsx output path
                wb.save(output_path)
                processed_files.append(output_path)
            else:
                # Multiple files -> Save to temp, later zipped
                temp_out = os.path.join(os.path.dirname(output_path), f"Processed_{base_fname}")
                wb.save(temp_out)
                processed_files.append(temp_out)
                
        # 3. Handle ZIP archiving if multiple files
        if len(opname_paths) > 1:
            with zipfile.ZipFile(output_path, 'w', zipfile.ZIP_DEFLATED) as zipf:
                for tf in processed_files:
                    zipf.write(tf, os.path.basename(tf))
                    # Cleanup the temporary modified file after zipping
                    os.remove(tf)

        return {"status": "success", "message": "Proses berhasil", "output": output_path, "files_processed": len(opname_paths)}

    except Exception as e:
        err_msg = traceback.format_exc()
        print(f"Error in recouncil processing: {err_msg}")
        return {"status": "error", "message": str(e)}

if __name__ == "__main__":
    import sys
    
    if len(sys.argv) < 2:
        print(json.dumps({"status": "error", "message": "Missing arguments"}))
        sys.exit(1)
        
    cmd = sys.argv[1]
    
    if cmd == "process":
        try:
            data_json = sys.argv[2]
            data = json.loads(data_json)
            
            opname_paths = data.get("opnames")
            master_path = data.get("master")
            output_path = data.get("output_path")
            
            result = run_recouncil(opname_paths, master_path, output_path)
            print(json.dumps(result))
            
        except Exception as e:
            print(json.dumps({"status": "error", "message": str(e)}))
